### PR TITLE
Add calibrate with jacobians for Cal3Bundler

### LINKED ISSUE #237: Mismatch between calibrate functions for Cal3_S2 and Cal3Bundler
## Description

`PinholeCamera` can accept both `Cal3_S2` and `Cal3Bundler` as calibrations, but the `calibrate` function is defined differently for each, such that when `camera.backproject(pn, depth)` is called, `Cal3Bundler` errors out.

## Steps to reproduce

This unit test when added to `testPinholeCamera.cpp` raises the error:

```cpp
TEST( PinholeCamera, Cal3Bundler) {
  Cal3Bundler calibration;
  Pose3 wTc;
  PinholeCamera<Cal3Bundler> camera(wTc, calibration);
  Point2 p(50, 100);
  Point3 point = camera.backproject(p, 10);
}
```

## Expected behavior

`PinholeCamera` (and primarily it's base class `PinholeBase`) should be able to call `calibration().calibrate` without any errors for all calibration types.

## Environment

N/A

### PR DESCRIPTION
**Pull request by varunagrawal**
Added an additional `calibrate` method for `Cal3Bundler` which compute jacobians.

Since the normal calibration method is iterative, I have approximated the jacobian as if a single iteration step has occurred, which should suffice assuming the projection is close to the ideal measurements.

I would definitely like a thorough review on this though.

Fixes #237

### CODE REVIEW

**Comment by varunagrawal** [comment]
Okay so this is breaking because calling code is getting confused which version of `calibrate` to call.

@dellaert should I merge the two `calibrate` functions into one?

**Comment by dellaert** [comment]
> Okay so this is breaking because calling code is getting confused which version of `calibrate` to call.
> 
> @dellaert should I merge the two `calibrate` functions into one?

Yes :-)

File: gtsam/geometry/Cal3Bundler.cpp [hunk 1]
```diff
@@ -118,6 +120,25 @@ Point2 Cal3Bundler::calibrate(const Point2& pi, const double tol) const {
     throw std::runtime_error(
         "Cal3Bundler::calibrate fails to converge. need a better initialization");
 
+  const double x = invKPi.x(), y = invKPi.y();
```
Comment (dellaert) [inline gtsam/geometry/Cal3Bundler.cpp:123]: remove invKPi variable, just init x and y on line 102, and use to init `pn` on line 105.

File: gtsam/geometry/Cal3Bundler.cpp [hunk 2]
```diff
@@ -118,6 +120,25 @@ Point2 Cal3Bundler::calibrate(const Point2& pi, const double tol) const {
...
+  const double xx = x * x, yy = y * y;
+  const double rr = xx + yy;
+  const double g = (1 + k1_ * rr + k2_ * rr * rr);
+  const double inv_f = 1 / f_, inv_g = 1 / g;
+
+  // Approximate the jacobians via a single iteration of g.
```
Comment (dellaert) [inline gtsam/geometry/Cal3Bundler.cpp:129]: Does this comment still hold? You are doing multiple iterations above

File: gtsam/geometry/Cal3Bundler.cpp [hunk 3]
```diff
@@ -118,6 +120,25 @@ Point2 Cal3Bundler::calibrate(const Point2& pi, const double tol) const {
...
+  if (Dcal) {
+    *Dcal << -inv_f * x, -x * inv_g * rr, -x * inv_g * rr * rr, -inv_f * y,
```
Comment (dellaert) [inline gtsam/geometry/Cal3Bundler.cpp:131]: Where does this come from? Document

File: gtsam/geometry/Cal3Bundler.cpp [hunk 4]
```diff
@@ -118,6 +120,25 @@ Point2 Cal3Bundler::calibrate(const Point2& pi, const double tol) const {
...
+        -y * inv_g * rr, -y * inv_g * rr * rr;
+  }
+  if (Dp) {
+    const double dg_du = (2 * k1_ * x * inv_f) + (4 * k2_ * rr * x * inv_f);
```
Comment (dellaert) [inline gtsam/geometry/Cal3Bundler.cpp:135]: same

File: gtsam/geometry/Cal3Bundler.cpp (hunk 2 above)
Comment (varunagrawal) [reply gtsam/geometry/Cal3Bundler.cpp:129]: We're doing multiple iterations but when calculating the jacobians, I assume only one iteration is done. So it neither the initial estimate or the final iterate (which could be after n<=10 iterations), but after a single iteration.

File: gtsam/geometry/Cal3Bundler.cpp (hunk 1 above)
Comment (varunagrawal) [reply gtsam/geometry/Cal3Bundler.cpp:123]: We still need invKPi in the iteration...

File: gtsam/geometry/Cal3Bundler.cpp (hunk 2 above)
Comment (varunagrawal) [reply gtsam/geometry/Cal3Bundler.cpp:129]: What would you recommend adopting? This was one of the reasons this PR took as long as it did.

Comment (dellaert) [reply gtsam/geometry/Cal3Bundler.cpp:129]: `calibrate` takes undistorted coordinates and makes them into pixels. This is iterative, but at end we have the correct relationship. Then we ask how the answer changes *differentially*, after we found this answer. So, it clearly needs to be the final iterate, as I understand it. So, I'd recommend taking rr, g etc out of the loop, initialize them from x and y, and then update at end of the loop so they are valid for the last estimate.

You still have to document how you got the derivatives. These things are more clear when there is a closed form solution, but it's not obvious when we are iterating on a nonlinear relationship.

File: gtsam/gtsam.i [hunk 5]
```diff
@@ -1053,7 +1053,7 @@ class Cal3_S2Stereo {
 class Cal3Bundler {
   // Standard Constructors
   Cal3Bundler();
-  Cal3Bundler(double fx, double k1, double k2, double u0, double v0);
+  Cal3Bundler(double fx, double k1, double k2, double u0, double v0, double tol);
```
Comment (dellaert) [inline gtsam/gtsam.i:1056]: Wrapper can’t do default arguments  so either undo or add another version.

File: gtsam/geometry/Cal3Bundler.cpp [hunk 6]
```diff
@@ -25,13 +25,13 @@ namespace gtsam {
...
-    f_(1), k1_(0), k2_(0), u0_(0), v0_(0) {
+    f_(1), k1_(0), k2_(0), u0_(0), v0_(0), tol_(1e-5) {
 }
 
 /* ************************************************************************* */
-Cal3Bundler::Cal3Bundler(double f, double k1, double k2, double u0, double v0) :
-    f_(f), k1_(k1), k2_(k2), u0_(u0), v0_(v0) {
-}
+Cal3Bundler::Cal3Bundler(double f, double k1, double k2, double u0, double v0,
+                         double tol)
```
Comment (dellaert) [inline gtsam/geometry/Cal3Bundler.cpp:33]: I like it

File: gtsam/geometry/Cal3Bundler.cpp [hunk 7]
```diff
@@ -94,21 +94,24 @@ Point2 Cal3Bundler::uncalibrate(const Point2& p, //
...
 /* ************************************************************************* */
-Point2 Cal3Bundler::calibrate(const Point2& pi, const double tol) const {
+Point2 Cal3Bundler::calibrate(const Point2& pi,
+                              OptionalJacobian<2, 3> Dcal,
+                              OptionalJacobian<2, 2> Dp) const {
   // Copied from Cal3DS2 :-(
   // but specialized with k1,k2 non-zero only and fx=fy and s=0
-  const Point2 invKPi((pi.x() - u0_)/f_, (pi.y() - v0_)/f_);
+  double x = (pi.x() - u0_)/f_, y = (pi.y() - v0_)/f_;
+  const Point2 invKPi(x, y);
```
Comment (dellaert) [inline gtsam/geometry/Cal3Bundler.cpp:103]: Not needed, x  ,y are never changed. Ensure this by making const. In fact add const to all things that never change

File: gtsam/geometry/Cal3Bundler.cpp [hunk 8]
```diff
@@ -118,6 +121,25 @@ Point2 Cal3Bundler::calibrate(const Point2& pi, const double tol) const {
     throw std::runtime_error(
         "Cal3Bundler::calibrate fails to converge. need a better initialization");
 
+  x = invKPi.x(), y = invKPi.y();
+  const double xx = x * x, yy = y * y;
```
Comment (dellaert) [inline gtsam/geometry/Cal3Bundler.cpp:125]: My suggestion was to take rr and g out of the loop above and make them non-const

File: gtsam/geometry/Cal3Bundler.cpp [hunk 9]
```diff
@@ -118,6 +121,25 @@ Point2 Cal3Bundler::calibrate(const Point2& pi, const double tol) const {
...
+  const double rr = xx + yy;
+  const double g = (1 + k1_ * rr + k2_ * rr * rr);
+  const double inv_f = 1 / f_, inv_g = 1 / g;
+
+  // Approximate the jacobians via a single iteration of g.
```
Comment (dellaert) [inline gtsam/geometry/Cal3Bundler.cpp:130]: Obsolete comment if you buy my argument above, but you are still using the initial estimate for jacobians?

**Comment by varunagrawal** [comment]
@dellaert great call on the Implicit Function theorem. Thinking in math was the right approach :smile: 
This is now ready for review.

+ @johnwlambert

File: gtsam/geometry/tests/testCal3Bundler.cpp [hunk 10]
```diff
@@ -69,11 +71,20 @@ TEST( Cal3Bundler, Duncalibrate)
   Matrix numerical2 = numericalDerivative22(uncalibrate_, K, p);
   CHECK(assert_equal(numerical1,Dcal,1e-7));
   CHECK(assert_equal(numerical2,Dp,1e-7));
-  CHECK(assert_equal(numerical1,K.D2d_calibration(p),1e-7));
-  CHECK(assert_equal(numerical2,K.D2d_intrinsic(p),1e-7));
-  Matrix Dcombined(2,5);
-  Dcombined << Dp, Dcal;
-  CHECK(assert_equal(Dcombined,K.D2d_intrinsic_calibration(p),1e-7));
```
Comment (varunagrawal) [inline gtsam/geometry/tests/testCal3Bundler.cpp:76]: These methods are deprecated, which is why I removed these tests.

File: gtsam/geometry/Cal3Bundler.cpp [hunk 11]
```diff
@@ -118,6 +121,25 @@ Point2 Cal3Bundler::calibrate(const Point2& pi, const double tol) const {
     throw std::runtime_error(
         "Cal3Bundler::calibrate fails to converge. need a better initialization");
 
+  // We make use of the Implicit Function Theorem to compute the Jacobians from uncalibrate
+  // Given f(pi, pn) = uncalibrate(pn) - pi, and g(pi) = calibrate, we can easily compute the Jacobians
+  // df/pi = -I (pn and pi are independent args)
+  // Dcal = -inv(H_uncal_pn) * df/pi = -inv(H_uncal_pn) * (-I) = inv(H_uncal_pn)
+  // Dp = -inv(H_uncal_pn) * df/K = -inv(H_uncal_pn) * H_uncal_K
+  Matrix H_uncal_K, H_uncal_pn;
```
Comment (dellaert) [inline gtsam/geometry/Cal3Bundler.cpp:129]: These matrices should be fixed-size. Also, cache H_uncal_pn here? (In a fixed size Matrix) finally, you can avoid calculating H_uncal_K using `Dcal ? &H_uncal_K : nullptr in first argument below.

**Review by dellaert — COMMENTED** [review]
Awesome. You *need* to use fixed-size matrices if you know sizes, esp in functions like these that could be called for all points.

Maybe make an issue to hunt down and fix other implicit theorem instances?

**Review by dellaert — APPROVED** [review]

**Comment by dellaert** [comment]
Approved as I’ll trust you’ll make those final changes

File: gtsam/geometry/Cal3Bundler.cpp (hunk 11 above)
Comment (varunagrawal) [reply gtsam/geometry/Cal3Bundler.cpp:129]: Quick question about caching H_uncal_pn: should I be calling uncalibrate with H_uncal_pn always?

**Pull request merged**